[export] grad of exported function fails with device assignment error

### Description

The `jax.experimental.export` exports the VJP using a synthetic mesh using the first N devices for the export platform. This seemed reasonable because all that is captured in the exported artifact is the number of devices, not their ids. However, during lowering there may be a conflict between the order of devices used by exporting code and shardings present in the primal functions.

For example, the following code fails:

```
    def f(x):
      return jnp.sum(x * 2.)

    mesh_rev = Mesh(list(reversed(jax.local_devices())), "i")
    shardings_rev = NamedSharding(mesh_rev, jax.sharding.PartitionSpec(("i",)))
    input_no_shards = jnp.ones(shape=(jax.local_device_count(),))
    input_rev = jax.device_put(input_no_shards, device=shardings_rev)
    exp_rev = export.export(pjit.pjit(f, in_shardings=shardings_rev))(input_no_shards)

    g = jax.grad(export.call(exp_rev))(input_rev)  # Failure here
```

The failure is:
```
E       ValueError: Received incompatible devices for pjitted computation. Got ARG_SHARDING with device ids [0, 1] on platform CPU and pjit inside pjit with device ids [1, 0] on platform CPU at /Users/necula/Source/jax/tests/export_test.py:1091:14 (JaxExportTest.test_grad_sharding_different_mesh)
```

Instead, the exporting of the VJP should use the same device assignment as for exporting the primal function.

### System info (python version, jaxlib version, accelerator, etc.)

Not needed

--- a/jax/experimental/export/_export.py
+++ b/jax/experimental/export/_export.py
@@ -351,6 +351,7 @@ def export(fun_jax: Callable,
            *,
            lowering_platforms: Sequence[str] | None = None,
            disabled_checks: Sequence[DisabledSafetyCheck] = (),
+           _device_assignment_for_internal_jax2tf_use_only = None,
            ) -> Callable[..., Exported]:
   """Exports native serialization for a JAX function.
 
@@ -413,12 +414,15 @@ def export(fun_jax: Callable,
         _experimental_lowering_parameters=mlir.LoweringParameters(
           platforms=actual_lowering_platforms,
         ))
-    return _export_lowered(lowered, disabled_checks=disabled_checks)
+    return _export_lowered(
+        lowered, disabled_checks=disabled_checks,
+        _device_assignment_for_internal_jax2tf_use_only=_device_assignment_for_internal_jax2tf_use_only)
   return do_export
 
 def _export_lowered(
     lowered: stages.Lowered,
     disabled_checks: Sequence[DisabledSafetyCheck] = (),
+    _device_assignment_for_internal_jax2tf_use_only = None,
   ) -> Exported:
   version = config.jax_serialization_version.value
   if (version < minimum_supported_serialization_version or
@@ -498,6 +502,8 @@ def _export_lowered(
     for s, aval in zip(lowering.compile_args["out_shardings"], out_avals_flat))
 
   device_assignment = lowering.compile_args["device_assignment"]
+  if _device_assignment_for_internal_jax2tf_use_only is not None:
+    _device_assignment_for_internal_jax2tf_use_only[0] = device_assignment
   def _get_exported_vjp(exp_primal: Exported) -> Exported:
     # Turn the primal jaxpr into a function, in preparation for exporting
     # the VJP. Note that jaxpr_as_fun produces a function with flat arguments

--- a/jax/experimental/jax2tf/jax2tf.py
+++ b/jax/experimental/jax2tf/jax2tf.py
@@ -514,11 +514,15 @@ class NativeSerializationImpl(SerializationImpl):
       _thread_local_state.call_tf_concrete_function_list = _prev_func_list
 
     self._restore_context = _restore_context
+    _exported_device_assignment = [None]
     self.exported = export.export(
         self.fun_jax,
         lowering_platforms=self.native_serialization_platforms,
-        disabled_checks=self.native_serialization_disabled_checks
+        disabled_checks=self.native_serialization_disabled_checks,
+        _device_assignment_for_internal_jax2tf_use_only=_exported_device_assignment,
     )(*self.args_specs, **self.kwargs_specs)
+    assert(_exported_device_assignment[0] is not None)
+    self.device_assignment = _exported_device_assignment[0]
 
   def after_conversion(self):
     self._restore_context()
@@ -531,15 +535,13 @@ class NativeSerializationImpl(SerializationImpl):
 
   def get_vjp_fun(self) -> tuple[Callable,
                                  Sequence[core.AbstractValue]]:
-    # TODO(necula): use the actual device assignment from the primal function
-    device_assignment = jax.devices(jax.default_backend())[:self.exported.nr_devices]
     return _export._get_vjp_fun(self.fun_jax,
                                 in_tree=self.exported.in_tree,
                                 in_avals=self.exported.in_avals,
                                 in_shardings=self.exported.in_shardings,
                                 out_avals=self.exported.out_avals,
                                 out_shardings=self.exported.out_shardings,
-                                device_assignment=device_assignment,
+                                device_assignment=self.device_assignment,
                                 apply_jit=True)
 
 class GraphSerializationImpl(SerializationImpl):

--- a/jax/experimental/jax2tf/tests/sharding_test.py
+++ b/jax/experimental/jax2tf/tests/sharding_test.py
@@ -455,22 +455,26 @@ class ShardingTest(tf_test_util.JaxToTfTestCase):
         ])
 
   def test_grad_sharding_different_mesh(self):
-    self.skipTest("TODO: fix the plumbing of device_assignment for jax2tf: https://github.com/google/jax/pull/21319")
     # Convert with two similar meshes, the only difference being
     # the order of the devices. grad should not fail.
     # https://github.com/google/jax/issues/21314
+    devices = jax.local_devices()[:2]
+    if len(devices) < 2:
+      raise unittest.SkipTest("Test requires 2 local devices")
     def f_jax(x):
       return jnp.sum(x * 2.)
 
-    mesh = Mesh(jax.local_devices(), "i")
+    mesh = Mesh(devices, "i")
     # The same mesh with reversed order of devices
-    mesh_rev = Mesh(list(reversed(jax.local_devices())), "i")
+    mesh_rev = Mesh(list(reversed(devices)), "i")
     shardings = NamedSharding(mesh, jax.sharding.PartitionSpec(("i",)))
     shardings_rev = NamedSharding(mesh_rev, jax.sharding.PartitionSpec(("i",)))
 
-    f_tf = tf.function(jax2tf.convert(pjit.pjit(f_jax, in_shardings=shardings)))
-    f_tf_rev = tf.function(jax2tf.convert(pjit.pjit(f_jax, in_shardings=shardings_rev)))
-    inp = np.ones((jax.local_device_count(), 4), dtype=np.float32)
+    f_tf = tf.function(jax2tf.convert(pjit.pjit(f_jax, in_shardings=shardings)),
+                       autograph=False)
+    f_tf_rev = tf.function(jax2tf.convert(pjit.pjit(f_jax, in_shardings=shardings_rev)),
+                           autograph=False)
+    inp = np.ones((2, 4), dtype=np.float32)
 
     input_v = tf.Variable(inp)
     with tf.GradientTape(persistent=True) as tape:

